agb/src/display/affine.rs

```diff
@@ -0,0 +1,223 @@
+#![warn(missing_docs)]
+//! # Affine matricies for the Game Boy Advance
+//!
+//! An affine matrix represents an affine transformation, an affine
+//! transformation being one which preserves parallel lines (note that this
+//! therefore cannot represent perspective seen in games like Super Mario Kart).
+//! Affine matricies are used in two places on the GBA, for affine backgrounds
+//! and for affine objects.
+//!
+//! # Linear Algebra basics
+//! As a matrix, they can be manipulated using linear algebra, although you
+//! shouldn't need to know linear algebra to use this apart from a few things
+//!
+//! If `A` and `B` are matricies, then matrix `C = A * B` represents the
+//! transformation `A` performed on `B`, or alternatively `C` is transformation
+//! `B` followed by transformation `A`.
+//!
+//! Additionally matrix multiplication is not commutative, meaning swapping the
+//! order changes the result, or `A * B ≢ B * A`.
+
+use core::{
+    convert::{TryFrom, TryInto},
+    ops::{Mul, MulAssign},
+};
+
+use agb_fixnum::{Num, Vector2D};
+
+type AffineMatrixElement = Num<i32, 8>;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+/// An affine matrix stored in a way that is efficient for the GBA to perform
+/// operations on. This implements multiplication.
+pub struct AffineMatrix {
+    a: AffineMatrixElement,
+    b: AffineMatrixElement,
+    c: AffineMatrixElement,
+    d: AffineMatrixElement,
+    x: AffineMatrixElement,
+    y: AffineMatrixElement,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+/// The error emitted upon a conversion that could not be performed due to
+/// overflowing the destination data size
+pub struct OverflowError(pub(crate) ());
+
+impl AffineMatrix {
+    #[must_use]
+    /// The Identity matrix. The identity matrix can be thought of as 1 and is
+    /// represented by `I`. For a matrix `A`, `A ≡ A * I ≡ I * A`.
+    pub fn identity() -> Self {
+        AffineMatrix {
+            a: 1.into(),
+            b: 0.into(),
+            c: 0.into(),
+            d: 1.into(),
+            x: 0.into(),
+            y: 0.into(),
+        }
+    }
+
+    #[must_use]
+    /// Generates the matrix that represents a rotation
+    pub fn from_rotation<const N: usize>(angle: Num<i32, N>) -> Self {
+        fn from_rotation(angle: Num<i32, 28>) -> AffineMatrix {
+            let cos = angle.cos().change_base();
+            let sin = angle.sin().change_base();
+
+            AffineMatrix {
+                a: cos,
+                b: sin,
+                c: -sin,
+                d: cos,
+                x: 0.into(),
+                y: 0.into(),
+            }
```

Comment:
maybe worth a comment saying this is backwards intentionally because gameboy advance?